Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp

```diff
@@ -2057,6 +2057,20 @@ ErrorOr<Optional<ReadonlyBytes>> JPEGImageDecoderPlugin::icc_data()
     return OptionalNone {};
 }
 
+Optional<Gps> JPEGImageDecoderPlugin::gps()
+{
+    if (m_context->exif_metadata && m_context->exif_metadata->gps_info().has_value()) {
+        auto gps_longitude = m_context->exif_metadata->gps_longitude();
+        double longitude = gps_longitude->at(0).as_double() + (gps_longitude->at(1).as_double() / 60.0) + (gps_longitude->at(2).as_double() / 3600.0);
+
+        auto gps_latitude = m_context->exif_metadata->gps_latitude();
+        double latitude = gps_latitude->at(0).as_double() + (gps_latitude->at(1).as_double() / 60.0) + (gps_latitude->at(2).as_double() / 3600.0);
+
+        return Gps { latitude, longitude };
+    }
+    return OptionalNone {};
+}
```

Comment:
We definitely need to find a way to factorize this code, we can't duplicate that in every image decoder. Also, operations on GPS coordinates doesn't seem to be a job for an image decoder.

The `Metadata` type was designed to be extended with more complex subtypes (e.g. `ExifMetadata`), but these subtypes can still be accessed by the same API.

So in `PropertiesWindow.cpp`, you can have something like:
```cpp
if (auto const& metadata = image_decoder->metadata(); metadata.has_value() && is<ExifMetadata>(*metadata)) {
	auto gps = gps_from_exif(metadata);
	...
}
```

---

Now on this code snippet specifically:

This is user provided data, you shouldn't assume that having `gps_info` means that all flags are present. You should still check both `metadata.gps_longitude()` and `metadata.gps_latitude()`.

Also, please use the `operator[]` instead of `at()`.

Reply:
Thanks for the review, I've moved the exif GPS parsing to a new class / file `ExifGPS.h`